x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/PrecompiledCharMapNormalizer.java

```diff
@@ -0,0 +1,203 @@
+     * <a href="https://github.com/google/sentencepiece/blob/bc53923a9147dc8ffa54034c8ed774de78cc4d39/third_party/darts_clone/darts.h#L469">
+     *     DARTS
+     *     </a>
+     * @param inputBytes utf8 bytes to normalize
+     * @param offset offset position to start given the input
+     * @param len the length of bytes to consider
+     * @return The starting position in the normalization string of the normalized bytes, if found.
+     */
+    OptionalInt commonPrefix(byte[] inputBytes, int offset, int len) {
+        int pos = 0;
+        OptionalInt vs = OptionalInt.empty();
+        int v = offsets[pos];
+        pos ^= offset(v);
+        for (int i = offset; i < offset + len; i++) {
+            // bytes can be negative in java, handle it and require unsigned
+            int k = inputBytes[i];
+            if (k < 0) {
+                k += 256;
+            }
+            if (k == 0) {
+                break;
+            }
+            pos ^= k;
+            v = offsets[pos];
+            if (label(v) != k) {
+                return vs;
+            }
+            pos ^= offset(v);
+            if (hasLeaf(v)) {
+                vs = OptionalInt.of(value(offsets[pos]));
+                return vs;
+            }
+        }
+        return vs;
+    }
+
+    Optional<BytesRef> normalizePart(byte[] strBytes, int offset, int len) {
+        OptionalInt index = commonPrefix(strBytes, offset, len);
+        if (index.isEmpty()) {
+            return Optional.empty();
+        }
+        int firstIndex = index.getAsInt();
+        int secondIndex = firstIndex;
+        // Parsed normalized string has normalization sections partitioned by \0 (NULL) byte
+        while (secondIndex < normalizedStrUtf8Bytes.length && normalizedStrUtf8Bytes[secondIndex] != 0) {
+            secondIndex++;
+        }
+        if (secondIndex == firstIndex) {
+            return Optional.empty();
+        }
+        return Optional.of(new BytesRef(normalizedStrUtf8Bytes, firstIndex, secondIndex - firstIndex));
+    }
+
+    String normalize(String str) {
+        // We need to iterate actual Unicode graphemes (this includes surrogate pairs, etc.)
+        // I would much rather translate the entire input string text into utf-8 bytes, and then iterate to the appropriate
+        // break points from there. But, this seemed the easiest way for now
+        BreakIterator b = BreakIterator.getCharacterInstance(Locale.ROOT);
+        b.setText(str);
+        int start = b.first();
+        // If we knew the utf-8 length ahead of time (and iterated over the bytes in the appropriate chunks)
+        // we could pre-populate the known length here.
+        BytesRefBuilder strBuilder = new BytesRefBuilder();
+        for (int end = b.next(); end != BreakIterator.DONE; start = end, end = b.next()) {
+            String unicodeStr = str.substring(start, end);
+            byte[] unicode = unicodeStr.getBytes(StandardCharsets.UTF_8);
+            if (unicode.length < 6) {
```

Comment:
I am not sure when the utf-8 length of a unicode grapheme would be 5 AND would be a surrogate pair. Additionally, I am not 100% sure the section of the code in the sentence piece normalizer that does this check. I will dig around to see. 

It very well may be "Hey, check if you are a surrogate pair, if so, split into chars and figure it out".

Reply:
Ah, OK, so the answer is that `BreakIterator.getCharacterInstance` does not simply split on Unicode codepoints when the locale is `Locale.ROOT`.

1575 and 1619 are not a surrogate pair (such values would be in the 50000s), but two separate characters that print as one.  1619 is a [non-spacing mark](https://unicodeplus.com/bidiclass/NSM)

And then `unicode.length < 6` is `true` for those two characters together.

In that case it will be much harder to simplify the code.  It might still be possible by iterating the UTF-16 characters but looking ahead to the next one and checking if it has a [combining class](https://www.compart.com/en/unicode/combining) that isn't "not reordered".

But I am happy to merge this with the current algorithm.